### PR TITLE
Signalbug and Ejectbug

### PR DESCRIPTION
**Pull request by AOhassan**
## Description

Issue # #1511,#1512 and #1456

Added signaling to is_provisioned and run methods.
created signaling file and functions to read/write signaling file.
implemented more logging for is_provisioned and run methods.
fixed regex syntax that caused telemetry tests to fail.
changed report_ready to return boolean.
deferred eject_dvd to be called after provisioning is complete.
implemented tests for eject_dvd.

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1528)

### CODE REVIEW

File: azurelinuxagent/pa/provision/default.py [hunk 1]
```diff
@@ -57,45 +58,71 @@ def run(self):
...
 
-            self.handle_provision_guest_agent(ovf_env.provision_guest_agent)
 
-            self.report_ready(thumbprint)
-            logger.info("Provisioning complete")
+                self.report_event("Provisioning succeeded ({0}s)".format(self._get_uptime_seconds()),
+                    is_success=True,
+                    duration=elapsed_milliseconds(utc_start))
+
+                self.handle_provision_guest_agent(ovf_env.provision_guest_agent)
```
Comment (jasonzio) [inline azurelinuxagent/pa/provision/default.py]: The very last thing that happens during provisioning should be the call to write_provisioned(); a crash immediately after that call would mean that, upon restart, provision_state == "provisioned_not_signaled" and the VM would immediately send the "done provisioning" signal to the control plane. However, as this is written, lines 110-117 are executed after write_provisioned(). Unless handle_provision_guest_agent() doesn't do any provisioning work (which I'm pretty sure is not a true statement), this code violates the expectation that provisioning is DONE when write_provisioned() is called.

File: azurelinuxagent/pa/provision/default.py [hunk 2]
```diff
@@ -166,11 +193,20 @@ def get_ssh_host_key_thumbprint(self, chk_err=True):
...
+        return os.path.isfile(self.provisioned_file_path())
+
+    def isfile_signaled(self):
+        return os.path.isfile(self.signaled_file_path())
+
     def is_provisioned(self):
         '''
-        A VM is considered provisionend *anytime* the provisioning
+        A VM is considered provisioned *anytime* the provisioning
         sentinel file exists and not provisioned *anytime* the file
```
Comment (jasonzio) [inline azurelinuxagent/pa/provision/default.py]: See my previous comment about handle_provision_guest_agent(). I don't think the code, as changed, lives up to this correct comment.

File: azurelinuxagent/pa/provision/default.py [hunk 3]
```diff
@@ -57,45 +58,68 @@ def run(self):
         if not conf.get_provision_enabled():
             logger.info("Provisioning is disabled, skipping.")
             self.write_provisioned()
-            self.report_ready()
+            logger.info("signaling...")
+            if self.report_ready():
+                self.write_signaled()
+                logger.info("signaling complete")
+            else:
+                logger.info("signaling incomplete")
```
Comment (Jdzions) [inline azurelinuxagent/pa/provision/default.py]: This block of code appears three times, and in all three cases subsequent code does not care which branch of the `if` was taken. You can hoist this up into a function local to the run() method.

File: azurelinuxagent/pa/provision/default.py (hunk 1 above)
Comment (Jdzions) [reply azurelinuxagent/pa/provision/default.py]: Turns out that handle_provision_guest_agent() does NOT do any provisioning work, just writes some log stuff. Weird name for a method that just logs.

I withdraw this objection.

---
In reply to: [284935536](https://github.com/Azure/WALinuxAgent/pull/1528#discussion_r284935536) [](ancestors = 284935536)

**Comment by narrieta** [comment]
@AOhassan - Hi, just checking if this PR is still active. Thanks!

**Comment by narrieta** [comment]
Please re-open if still active. Thanks!

**Pull request closed without merging**